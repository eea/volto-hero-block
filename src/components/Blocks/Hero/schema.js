import { addStyling } from '@plone/volto/helpers';
import config from '@plone/volto/registry';

import alignTopSVG from '@plone/volto/icons/move-up.svg';
import alignCenterSVG from '@plone/volto/icons/row.svg';
import alignBottomSVG from '@plone/volto/icons/move-down.svg';

const ALIGN_INFO_MAP = {
  'has--bg--top': [alignTopSVG, 'Top'],
  'has--bg--center': [alignCenterSVG, 'Center'],
  'has--bg--bottom': [alignBottomSVG, 'Bottom'],
};

export const HeroBlockSchema = () => {
  return {
    title: 'Hero',
    fieldsets: [
      {
        id: 'default',
        title: 'Default',
        fields: [
          'fullWidth',
          'fullHeight',
          'quoted',
          'spaced',
          'inverted',
          'buttonLabel',
          'buttonLink',
          'overlay',
          'image',
        ],
      },
      {
        id: 'copyright',
        title: 'Copyright',
        fields: ['copyright', 'copyrightIcon', 'copyrightPosition'],
      },
    ],
    properties: {
      fullWidth: {
        title: 'Full width',
        type: 'boolean',
        defaultValue: true,
      },
      fullHeight: {
        title: 'Full height',
        type: 'boolean',
        defaultValue: true,
      },
      quoted: {
        title: 'Quoted',
        type: 'boolean',
        defaultValue: false,
      },
      spaced: {
        title: 'Spaced',
        type: 'boolean',
        defaultValue: false,
      },
      inverted: {
        title: 'Inverted',
        type: 'boolean',
        defaultValue: true,
      },
      buttonLabel: {
        title: 'Button label',
        widget: 'textarea',
      },
      buttonLink: {
        title: 'Button link',
        widget: 'url',
      },
      overlay: {
        title: 'Image darken overlay',
        type: 'boolean',
        defaultValue: true,
      },
      image: {
        title: 'Image',
        widget: 'attachedimage',
      },
      copyright: {
        title: 'Text',
      },
      copyrightIcon: {
        title: 'Icon',
        description: (
          <>
            Ex. ri-copyright-line. See{' '}
            <a
              target="_blank"
              rel="noopener noreferrer"
              href="https://remixicon.com/"
            >
              Remix Icon set
            </a>
          </>
        ),
        default: 'ri-copyright-line',
      },
      copyrightPosition: {
        title: 'Align',
        widget: 'align',
        actions: ['left', 'right'],
        defaultValue: 'left',
      },
    },
    required: [],
  };
};

export const stylingSchema = (props) => {
  const schema = addStyling(props);
  schema.properties.styles.schema = {
    title: 'Hero style',
    block: 'hero',
    fieldsets: [
      {
        id: 'default',
        title: 'Default',
        fields: [
          'backgroundVariant',
          'bg',
          'alignContent',
          'textAlign',
          'textVariant',
          'buttonVariant',
          'buttonAlign',
        ],
      },
    ],
    properties: {
      bg: {
        title: 'Background image position',
        widget: 'align',
        actions: Object.keys(ALIGN_INFO_MAP),
        actionsInfoMap: ALIGN_INFO_MAP,
        defaultValue: 'has--bg--center',
      },
      backgroundVariant: {
        title: 'Background theme',
        widget: 'theme_picker',
        colors: [
          ...(config.settings && config.settings.themeColors
            ? config.settings.themeColors.map(({ value, title }) => ({
                name: value,
                label: title,
              }))
            : []),
          //and add extra ones here
        ],
      },
      alignContent: {
        title: 'Content align',
        choices: [
          ['start', 'Top'],
          ['center', 'Center'],
          ['end', 'Bottom'],
        ],
      },
      textAlign: {
        title: 'Text align',
        widget: 'align',
        actions: ['left', 'center', 'right'],
        defaultValue: 'left',
      },
      textVariant: {
        title: 'Text theme',
        widget: 'theme_picker',
        colors: [
          ...(config.settings && config.settings.themeColors
            ? config.settings.themeColors.map(({ value, title }) => ({
                name: value,
                label: title,
              }))
            : []),
          //and add extra ones here
        ],
      },
      buttonVariant: {
        title: 'Button theme',
        widget: 'theme_picker',
        colors: [
          ...(config.settings && config.settings.themeColors
            ? config.settings.themeColors.map(({ value, title }) => ({
                name: value,
                label: title,
              }))
            : []),
          //and add extra ones here
        ],
      },
      buttonAlign: {
        title: 'Button align',
        widget: 'align',
        actions: ['left', 'center', 'right'],
        defaultValue: 'left',
      },
    },
    required: [],
  };
  return schema;
};
