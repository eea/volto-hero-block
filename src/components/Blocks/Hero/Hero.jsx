import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import { isInternalURL } from '@plone/volto/helpers/Url/Url';
import { isImageGif } from '../../../helpers';

Hero.propTypes = {
  image: PropTypes.string,
  fullWidth: PropTypes.bool,
  fullHeight: PropTypes.bool,
  alignContent: PropTypes.string,
  textAlign: PropTypes.string,
  backgroundVariant: PropTypes.string,
  quoted: PropTypes.bool,
  spaced: PropTypes.bool,
  inverted: PropTypes.bool,
  textVariant: PropTypes.string,
};

function Hero({
  image,
  overlay = true,
  fullWidth = true,
  fullHeight = true,
  children,
  spaced = false,
  inverted = true,
  styles,
}) {
  const isExternal = !isInternalURL(image);
  const { alignContent = 'center', backgroundVariant = 'primary' } =
    styles || {};
  return (
    <div
      className={cx(
        'eea hero-block',
        !image &&
          backgroundVariant &&
          !fullWidth &&
          `color-bg-${backgroundVariant}`,
        {
          spaced,
          inverted,
          'full-height': fullHeight,
        },
      )}
    >
      <div
        className={cx(
          'hero-block-image-wrapper',
          !image &&
            backgroundVariant &&
            fullWidth &&
            `color-bg-${backgroundVariant}`,
          {
            'full-width': fullWidth,
          },
        )}
      >
        <div
          className={cx('hero-block-image')}
          style={
            image
              ? {
                  backgroundImage: isExternal
                    ? `url(${image})`
                    : isImageGif(image)
                    ? `url(${image}/@@images/image)`
                    : `url(${image}/@@images/image/huge)`,
                }
              : {}
          }
        />
        {image && overlay && (
          <div className="hero-block-image-overlay dark-overlay-4"></div>
        )}
      </div>
      <div
        className={cx(
          'hero-block-inner-wrapper d-flex',
          `flex-items-${alignContent}`,
        )}
      >
        <div className="hero-block-body">{children}</div>
      </div>
    </div>
  );
}

Hero.Text = ({ children, quoted, styles }) => {
  const { textVariant = 'white', textAlign = 'left' } = styles || {};
  return (
    <div
      className={cx(
        'hero-block-text',
        `color-fg-${textVariant}`,
        `text-${textAlign}`,
      )}
    >
      <div className={quoted ? 'quoted-wrapper' : ''}>{children}</div>
    </div>
  );
};

Hero.Meta = ({ children, styles }) => {
  const { buttonAlign = 'left' } = styles || {};
  return (
    <div className={cx('hero-block-meta', `text-${buttonAlign}`)}>
      {children}
    </div>
  );
};

export default Hero;
