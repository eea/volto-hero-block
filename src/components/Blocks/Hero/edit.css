.hero-block .block-add-button,
.hero-block .drag.handle.wrapper,
.hero-block .block-editor-slate:first-of-type .delete-button {
  display: none !important;
}
