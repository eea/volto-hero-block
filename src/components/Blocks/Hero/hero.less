@type: 'extra';
@element: 'hero';

@import (multiple) './hero.variables';
@import (multiple) '@plone/volto/theme/themes/default/globals/site.variables';
& {
  @import (less) 'remixicon/fonts/remixicon.css';
}
/*-------------------
      Hero
--------------------*/

.block.hero .ui.delete-button {
    color: @heroDeleteButtonColor !important;
}

.eea.hero-block {
  position: @heroBlockPosition;
}

.eea.hero-block.spaced {
  margin: @heroBlockSpacedMargin;
}

.hero-block-inner-wrapper {
  height: 100%;
  padding-block: @heroBlockWrapperPadding;
  position: relative;
}

.hero-block-body {
  display: flex;
  width: 100%;
  flex-direction: column;
  gap: @mobileHeroBlockBodyGap;
}

.hero-block-image {
  background-color: inherit;
  background-repeat: @heroBlockImageBackgroundRepeat;
  background-size: @heroBlockImageBackgroundSize;
  background-position: @heroBlockImageBackgroundPosition;
  width: 100%;
  height: 100%;
  position: absolute;
  z-index: 0;
}

.hero-block-image-wrapper {
  position: absolute;
  min-height: @heroBlockImageWrapperMinHeight;
  width: 100%;
  height: 100%;
}

.hero-block-body .eea.copyright {
  position: absolute;
  bottom: @mobileCopyrightBottom;
}

.hero-block-text .quoted > * {
  display: inline-block;
}

.hero-block[class*='color-bg']:not(.full-width) {
  padding: @heroBlockHomePagePadding;
}

.eea.hero-block.full-height,
.eea.hero-block.full-height .hero-block-image-wrapper,
.eea.hero-block.full-height .hero-block-inner-wrapper {
  min-height: @mobileHeroBlockMinHeight;
}

@media only screen and (min-width: @tabletBreakpoint) {
  .eea.hero-block.full-height,
  .eea.hero-block.full-height .hero-block-image-wrapper,
  .eea.hero-block.full-height .hero-block-inner-wrapper {
    min-height: @tabletHeroBlockMinHeight;
  }
  .hero-block-body {
    gap: @tabletHeroBlockBodyGap;
  }
  .hero-block-inner-wrapper {
    padding-block: @tableHeroBlockWrapperPadding;
  }
  .homepage #page-document > .eea.hero-block:first-of-type .hero-block-body {
    margin: @heroBlockHomePageBodyMargin;
  }
}

@media only screen and (min-width: @computerBreakpoint) {
  .eea.hero-block.full-height,
  .eea.hero-block.full-height .hero-block-image-wrapper,
  .eea.hero-block.full-height .hero-block-inner-wrapper {
    min-height: @desktopHeroBlockMinHeight;
  }

  .hero-block-inner-wrapper {
    padding-block: @desktopHeroBlockWrapperPadding;
  }
  .hero-block-body {
    gap: @desktopHeroBlockBodyGap;
  }
  .hero-block-body .eea.copyright {
    bottom: @desktopCopyrightBottom;
  }
}

@media print {
  .eea.hero-block.full-height,
  .eea.hero-block.full-height .hero-block-image-wrapper,
  .eea.hero-block.full-height .hero-block-inner-wrapper {
    min-height: unset;
  }

  .hero-block-inner-wrapper {
    padding-top: @heroBlockPrintPadding;
    padding-bottom: @heroBlockPrintPadding;
  }

  .hero-block-image-wrapper {
    left: 0 !important;
    right: 0 !important;
    padding-left: 0 !important;
    margin-left: 0 !important;
    margin-right: 0 !important;
    padding-right: 0% !important;
    width: 100% !important;
    break-inside: avoid-page;
  }

  .hero-block {
    break-inside: avoid-page;
  }

  .hero-block-body .eea.copyright {
    display: none;
    visibility: hidden;
  }
}

#page-document > .styled-hero {
  margin-top: 0;
}

.full-width1 {
  --sidebar-width: 0px;
  --toolbar-width: 80px;
  @toolbarWidthCSS: var(--toolbar-width, 0px);
  @sidebarWidthCSS: var(--sidebar-width, 0px);
  @scrollbarWidthCSS: var(--scrollbar-width, @scrollbarWidth);
  position: relative;
  right: 50%;
  left: 50%;
  width: calc(
    100vw - @toolbarWidthCSS - @sidebarWidthCSS - @scrollbarWidthCSS
  ) !important;
  max-width: initial !important;
  margin-right: calc(
    -50vw + (@toolbarWidthCSS + @sidebarWidthCSS + @scrollbarWidthCSS) / 2
  ) !important;
  margin-left: calc(
    -50vw + (@toolbarWidthCSS + @sidebarWidthCSS + @scrollbarWidthCSS) / 2
  ) !important;
}
.color-bg-primary { background-color: @primaryColor !important; }
.color-bg-secondary { background-color: @brown !important; }
// .color-bg-tertiary { background-color: @hoverBackgroundColor !important; }
.dark-overlay-4 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background: #000;
  opacity: 0.4;
  z-index: 0;
}
.text-left {
  text-align: left;
}
.text-center{
  text-align: center;
}
.text-right {
  text-align: right;
}
.d-flex { display: flex !important; }
.flex-row { flex-direction: row !important; }
.flex-items-start { align-items: flex-start !important; }
.flex-items-end { align-items: flex-end !important; }
.flex-items-center { align-items: center !important; }

.has--bg--top {
  background-position: top !important;
}

.has--bg--center {
  background-position: center !important;
}

.has--bg--bottom {
  background-position: bottom !important;
}

.eea.copyright {
  width: @copyrightWidth;

  .wrapper {
      display: flex;
      // color: @copyrightContentColor;
      // font-size: @copyrightContentFontSize;
      // align-items: @copyrightAlignItems;
      // line-height: @copyrightLineHeight;
      // gap: @copyrightContentPaddingRight;
      justify-content: var(--justify-content, flex-start);
  }
&.left .wrapper {
  --justify-content: flex-start;
}
&.center .wrapper {
  --justify-content: center;
}
&.right .wrapper {
  --justify-content: flex-end;
}
.icon {
  margin: 0; // override default
}
}

.quoted-wrapper h2::before,
.quoted-wrapper h3::before,
.quoted::before {
  content: '\EC51';
  font-family: 'remixicon' !important;

}

.quoted-wrapper  h2::after,
.quoted-wrapper  h3::after,
.quoted::after {
  content: "\ec52";
  font-family: 'remixicon' !important;
}
