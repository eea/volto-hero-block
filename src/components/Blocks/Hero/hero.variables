/* ==========================================================================
   Breakpoints
   ========================================================================== */
@tabletBreakpoint : 768px;
@computerBreakpoint : 992px;

/* ==========================================================================
   Base Sizes and Units
   ========================================================================== */
@emSize : 14px;

/* Rem Spacing Scale (based on 14px em size) */
@rem-space-1     : @4px; // 4px
@rem-space-3     : @12px; // 12px
@rem-space-4     : unit((16 / @emSize), rem); // 16px
@rem-space-5     : unit((20 / @emSize), rem); // 20px
@rem-space-6     : unit((24 / @emSize), rem); // 24px
@rem-space-7-50  : unit((30 / @emSize), rem); // 30px
@rem-space-8     : unit((32 / @emSize), rem); // 32px
@rem-space-10    : unit((40 / @emSize), rem); // 40px
@rem-space-12    : unit((48 / @emSize), rem); // 48px
@rem-space-20    : unit((80 / @emSize), rem); // 80px

/* ==========================================================================
   Hero Block - General
   ========================================================================== */
@heroBlockPosition : relative;
@heroDeleteButtonColor : white;
@scrollbarWidth : 15px;

/* ==========================================================================
   Hero Block - Min Heights
   ========================================================================== */
@mobileHeroBlockMinHeight  : 412px;
@tabletHeroBlockMinHeight  : 446px;
@desktopHeroBlockMinHeight : 664px;

/* ==========================================================================
   Hero Block - Padding and Margins
   ========================================================================== */
/* Homepage Specific */
@heroBlockHomePagePadding    : 0 @rem-space-5;
@heroBlockHomePageBodyMargin : @rem-space-20 0 0 0;

/* Spaced Version */
@heroBlockSpacedMargin : @rem-space-12 0;

/* Wrapper Padding */
@heroBlockWrapperPadding         : 5rem;
@tableHeroBlockWrapperPadding    : 6rem;
@desktopHeroBlockWrapperPadding  : 194px 96px;
@heroBlockPrintPadding           : 20px;

/* ==========================================================================
   Hero Block - Body Gap
   ========================================================================== */
@mobileHeroBlockBodyGap  : @rem-space-6;
@tabletHeroBlockBodyGap  : @rem-space-8;
@desktopHeroBlockBodyGap : @rem-space-10;

/* ==========================================================================
   Hero Block - Image
   ========================================================================== */
/* Image Wrapper */
@heroBlockImageWrapperMinHeight : 290px; // 5rem gap plus one line text

/* Background Image Styling */
@heroBlockImageBackgroundRepeat   : no-repeat;
@heroBlockImageBackgroundSize     : cover;
@heroBlockImageBackgroundPosition : center;

/* ==========================================================================
   Copyright Section
   ========================================================================== */
@copyrightWidth : 100%;

@mobileCopyrightBottom  : @rem-space-4;
@desktopCopyrightBottom : @rem-space-7-50;

@copyrightContentColor        : var(--text-color, #F9F9F9;);
@copyrightContentFontSize     : .75rem;
@copyrightLineHeight          : 1.2;
@copyrightContentPaddingRight : @rem-space-1;

@copyrightWrapperPaddingLeft  : @rem-space-3;
@copyrightWrapperPaddingRight : @rem-space-3;
@copyrightWrapperBottom       : @rem-space-4;
@copyrightAlignItems          : start;
