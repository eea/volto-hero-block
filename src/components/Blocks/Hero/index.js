import codeSVG from '@plone/volto/icons/code.svg';
import HeroEdit from './Edit';
import HeroView from './View';
import { stylingSchema } from './schema';
import LayoutSchema from './LayoutSchema';
import { defineMessages, createIntlCache, createIntl } from 'react-intl';

const messages = defineMessages({
  heroTitle: {
    id: 'Hero',
    defaultMessage: 'Hero',
  },
});

const cache = createIntlCache();

const intl = createIntl(
  {
    locale: 'en',
    messages: messages,
  },
  cache,
);

const applyConfig = (config) => {
  config.blocks.blocksConfig.hero = {
    id: 'hero',
    title: 'Hero',
    icon: codeSVG,
    group: 'common',
    edit: HeroEdit,
    view: HeroView,
    blockHasOwnFocusManagement: true,
    schemaEnhancer: stylingSchema,
    restricted: false,
    mostUsed: false,
    sidebarTab: 1,
    copyrightPrefix: '',
    security: {
      addPermission: [],
      view: [],
    },
    schema: LayoutSchema(intl),
  };

  config.settings.blocksWithFootnotesSupport = {
    ...(config.settings.blocksWithFootnotesSupport || {}),
    hero: ['text'],
  };

  return config;
};

export default applyConfig;
