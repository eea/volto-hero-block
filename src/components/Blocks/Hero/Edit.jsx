import React, { useState } from 'react';
import cx from 'classnames';
import isFunction from 'lodash/isFunction';
import { Icon } from 'semantic-ui-react';
import config from '@plone/volto/registry';
import { BlocksForm } from '@plone/volto/components';
import EditBlockWrapper from '@plone/volto/components/manage/Blocks/Block/EditBlockWrapper';
import { v4 as uuid } from 'uuid';
import '@eeacms/volto-hero-block/components/Blocks/Hero/edit.css';

import { isEmpty } from 'lodash';
import {
  BlockDataForm,
  SidebarPortal,
  UniversalLink,
} from '@plone/volto/components';
import { BodyClass } from '@plone/volto/helpers';

import { getFieldURL } from '@eeacms/volto-hero-block/helpers';
import { HeroBlockSchema } from './schema';
import Copyright from './Copyright';
import Hero from './Hero';

const Metadata = ({ buttonLabel, inverted, styles, ...props }) => {
  const buttonLink = getFieldURL(props.buttonLink);
  const { buttonVariant } = styles || {};

  return buttonLabel ? (
    <UniversalLink
      className={cx('ui button', buttonVariant, { inverted })}
      href={buttonLink || ''}
    >
      {buttonLabel}
    </UniversalLink>
  ) : (
    ''
  );
};

export default function Edit(props) {
  const id = uuid();
  const [selectedBlock, setSelectedBlock] = useState(id);
  const {
    data = {},
    block = null,
    selected,
    properties,
    onChangeBlock,
    onChangeField,
    pathname,
    metadata = null,
    setSidebarTab,
  } = props;
  const { copyright, copyrightIcon, copyrightPosition } = data;

  const copyrightPrefix = config.blocks.blocksConfig.hero.copyrightPrefix || '';
  const schema = React.useMemo(() => {
    if (isFunction(HeroBlockSchema)) {
      return HeroBlockSchema(props);
    }
    return HeroBlockSchema;
  }, [props]);

  const blockState = {};
  const data_blocks = data?.data?.blocks;

  if (data?.text || isEmpty(data_blocks)) {
    let dataWithoutText = { ...data };
    if (dataWithoutText) delete dataWithoutText.text;

    onChangeBlock(block, {
      ...dataWithoutText,
      data: data?.text
        ? {
            blocks: {
              [id]: {
                '@type': 'slate',
                value: data.text,
                plaintext: data.text?.[0].children?.[0].text,
              },
            },
            blocks_layout: { items: [id] },
          }
        : {
            blocks: {
              [id]: {
                '@type': 'slate',
                value: [{ type: 'h2', children: [{ text: '' }] }],
                plaintext: '',
              },
            },
            blocks_layout: { items: [id] },
          },
    });
  }

  return (
    <>
      <BodyClass className="with-hero-block" />
      <div
        className="hero-edit-wrapper"
        role="presentation"
        onClick={(e) => {
          const className = e.target.getAttribute('class');
          if (className && className.includes('hero')) {
            setSelectedBlock(id);
            setSidebarTab(1);
          }
        }}
      >
        <Hero {...data}>
          <Hero.Text {...data}>
            <BlocksForm
              metadata={properties || metadata}
              properties={data.data || {}}
              manage={false}
              allowedBlocks={'slate'}
              selectedBlock={selectedBlock}
              title={data.placeholder}
              onSelectBlock={(s, e) => {
                setSelectedBlock(s);
              }}
              onChangeFormData={(newFormData) => {
                onChangeBlock(block, {
                  ...data,
                  data: newFormData,
                });
              }}
              onChangeField={(id, value) => {
                if (['blocks', 'blocks_layout'].indexOf(id) > -1) {
                  blockState[id] = value;
                  onChangeBlock(block, {
                    ...data,
                    data: {
                      ...data.data,
                      ...blockState,
                    },
                  });
                } else {
                  onChangeField(id, value);
                }
              }}
              pathname={pathname}
            >
              {({ draginfo }, editBlock, blockProps) => (
                <EditBlockWrapper
                  draginfo={draginfo}
                  blockProps={blockProps}
                  disabled={data.disableInnerButtons}
                >
                  {editBlock}
                </EditBlockWrapper>
              )}
            </BlocksForm>
          </Hero.Text>
          <Hero.Meta {...data}>
            <Metadata {...data} />
          </Hero.Meta>
          {copyright ? (
            <Copyright copyrightPosition={copyrightPosition}>
              <Copyright.Prefix>{copyrightPrefix}</Copyright.Prefix>
              <Copyright.Icon>
                <Icon className={copyrightIcon} />
              </Copyright.Icon>
              <Copyright.Text>{copyright}</Copyright.Text>
            </Copyright>
          ) : (
            ''
          )}
        </Hero>
      </div>
      <SidebarPortal selected={selected}>
        <BlockDataForm
          block={block}
          schema={schema}
          title={schema.title}
          onChangeBlock={props.onChangeBlock}
          onChangeField={(id, value) => {
            onChangeBlock(block, {
              ...data,
              [id]: value,
            });
          }}
          formData={data}
        />
      </SidebarPortal>
    </>
  );
}
