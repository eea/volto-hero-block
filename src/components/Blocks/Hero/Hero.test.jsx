import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-intl-redux';
import configureStore from 'redux-mock-store';
import config from '@plone/volto/registry';
import '@testing-library/jest-dom/extend-expect';

import Hero from './Hero';

const { settings } = config;

const mockStore = configureStore();

const observe = jest.fn();
const unobserve = jest.fn();
window.IntersectionObserver = jest.fn((callback) => ({
  observe,
  unobserve,
}));

jest.mock('@eeacms/volto-hero-block/hooks', () => ({
  useFirstVisited: jest.fn(() => true),
}));

describe('Hero block', () => {
  it('renders a hero component', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const { container } = render(
      <Provider store={store}>
        <Hero
          image=""
          overlay={true}
          fullWidth={true}
          fullHeight={true}
          spaced={false}
          inverted={true}
          styles={{ alignContent: 'center', backgroundVariant: 'primary' }}
        >
          <Hero.Text
            quoted={false}
            styles={{ textVariant: 'white', textAlign: 'left' }}
          >
            Text test
          </Hero.Text>
          <Hero.Meta styles={{ buttonAlign: 'left' }}>Test meta</Hero.Meta>
        </Hero>
      </Provider>,
    );

    expect(container.querySelector('.eea.hero-block')).toBeInTheDocument();
    expect(container.querySelector('.inverted')).toBeInTheDocument();
    expect(container.querySelector('.full-height')).toBeInTheDocument();
    expect(container.querySelector('.full-width')).toBeInTheDocument();
  });

  it('renders a hero component', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const { container } = render(
      <Provider store={store}>
        <Hero
          image=""
          overlay={true}
          fullWidth={false}
          fullHeight={true}
          spaced={false}
          inverted={true}
          styles={{ alignContent: 'center', backgroundVariant: 'primary' }}
        >
          <Hero.Text
            quoted={false}
            styles={{ textVariant: 'white', textAlign: 'left' }}
          >
            Text test
          </Hero.Text>
          <Hero.Meta styles={{ buttonAlign: 'left' }}>Test meta</Hero.Meta>
        </Hero>
      </Provider>,
    );

    expect(container.querySelector('.eea.hero-block')).toBeInTheDocument();
    expect(container.querySelector('.inverted')).toBeInTheDocument();
    expect(container.querySelector('.full-height')).toBeInTheDocument();
    expect(container.querySelector('.full-width')).not.toBeInTheDocument();
  });

  it('renders a hero component', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const { container } = render(
      <Provider store={store}>
        <Hero
          image={{ '@type': 'URL', url: 'url_url', href: 'href_url' }}
          overlay={true}
          fullWidth={false}
          fullHeight={true}
          spaced={false}
          inverted={true}
          styles={{ alignContent: 'center', backgroundVariant: 'primary' }}
        >
          <Hero.Text
            quoted={false}
            styles={{ textVariant: 'white', textAlign: 'left' }}
          >
            Text test
          </Hero.Text>
          <Hero.Meta styles={{ buttonAlign: 'left' }}>Test meta</Hero.Meta>
        </Hero>
      </Provider>,
    );

    expect(container.querySelector('.eea.hero-block')).toBeInTheDocument();
    expect(container.querySelector('.hero-block-image')).toHaveStyle({
      backgroundImage: 'url(url_url)',
    });
  });

  it('renders a hero component', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const { container } = render(
      <Provider store={store}>
        <Hero
          image={`${settings.apiPath}/foo/bar`}
          overlay={true}
          fullWidth={false}
          fullHeight={true}
          spaced={false}
          inverted={true}
          styles={{ alignContent: 'center', backgroundVariant: 'primary' }}
        >
          <Hero.Text
            quoted={false}
            styles={{ textVariant: 'white', textAlign: 'left' }}
          >
            Text test
          </Hero.Text>
          <Hero.Meta styles={{ buttonAlign: 'left' }}>Test meta</Hero.Meta>
        </Hero>
      </Provider>,
    );

    expect(container.querySelector('.eea.hero-block')).toBeInTheDocument();
    expect(container.querySelector('.hero-block-image')).toHaveStyle({
      backgroundImage: 'url(/foo/bar/@@images/image/huge)',
    });
  });

  it('renders a hero component', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const { container } = render(
      <Provider store={store}>
        <Hero
          image={`${settings.apiPath}/foo/bar.gif`}
          styles={{ alignContent: undefined, backgroundVariant: undefined }}
        >
          <Hero.Text
            quoted={true}
            styles={{ textVariant: undefined, textAlign: undefined }}
          >
            Text test
          </Hero.Text>
          <Hero.Meta styles={{ buttonAlign: undefined }}>Test meta</Hero.Meta>
        </Hero>
      </Provider>,
    );

    expect(container.querySelector('.quoted-wrapper')).toBeInTheDocument();
  });

  it('renders a hero component', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const { container } = render(
      <Provider store={store}>
        <Hero image={`${settings.apiPath}/foo/bar.gif`} styles={undefined}>
          <Hero.Text quoted={true} styles={undefined}>
            Text test
          </Hero.Text>
          <Hero.Meta styles={undefined}>Test meta</Hero.Meta>
        </Hero>
      </Provider>,
    );

    expect(container.querySelector('.quoted-wrapper')).toBeInTheDocument();
  });
});
