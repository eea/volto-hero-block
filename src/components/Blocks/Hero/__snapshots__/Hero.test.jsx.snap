// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero block renders a hero component 1`] = `
<div
  className="eea hero-block inverted full-height"
>
  <div
    className="hero-block-image-wrapper color-bg-primary full-width"
  >
    <div
      className="hero-block-image"
      style={Object {}}
    />
    
  </div>
  <div
    className="hero-block-inner-wrapper d-flex flex-items-center"
  >
    <div
      className="hero-block-body"
    >
      <div
        className="hero-block-text color-fg-white text-left"
      >
        <div
          className=""
        >
          Text test
        </div>
      </div>
      <div
        className="hero-block-meta text-left"
      >
        Test meta
      </div>
    </div>
  </div>
</div>
`;

exports[`Hero block renders a hero component 2`] = `
<div
  className="eea hero-block color-bg-primary inverted full-height"
>
  <div
    className="hero-block-image-wrapper"
  >
    <div
      className="hero-block-image"
      style={Object {}}
    />
    
  </div>
  <div
    className="hero-block-inner-wrapper d-flex flex-items-center"
  >
    <div
      className="hero-block-body"
    >
      <div
        className="hero-block-text color-fg-white text-left"
      >
        <div
          className=""
        >
          Text test
        </div>
      </div>
      <div
        className="hero-block-meta text-left"
      >
        Test meta
      </div>
    </div>
  </div>
</div>
`;

exports[`Hero block renders a hero component 3`] = `
<div
  className="eea hero-block inverted full-height"
>
  <div
    className="hero-block-image-wrapper"
  >
    <div
      className="hero-block-image"
      style={
        Object {
          "backgroundImage": "url(url_url)",
        }
      }
    />
    <div
      className="hero-block-image-overlay dark-overlay-4"
    />
  </div>
  <div
    className="hero-block-inner-wrapper d-flex flex-items-center"
  >
    <div
      className="hero-block-body"
    >
      <div
        className="hero-block-text color-fg-white text-left"
      >
        <div
          className=""
        >
          Text test
        </div>
      </div>
      <div
        className="hero-block-meta text-left"
      >
        Test meta
      </div>
    </div>
  </div>
</div>
`;

exports[`Hero block renders a hero component 4`] = `
<div
  className="eea hero-block inverted full-height"
>
  <div
    className="hero-block-image-wrapper"
  >
    <div
      className="hero-block-image"
      style={
        Object {
          "backgroundImage": "url(/foo/bar/@@images/image/huge)",
        }
      }
    />
    <div
      className="hero-block-image-overlay dark-overlay-4"
    />
  </div>
  <div
    className="hero-block-inner-wrapper d-flex flex-items-center"
  >
    <div
      className="hero-block-body"
    >
      <div
        className="hero-block-text color-fg-white text-left"
      >
        <div
          className=""
        >
          Text test
        </div>
      </div>
      <div
        className="hero-block-meta text-left"
      >
        Test meta
      </div>
    </div>
  </div>
</div>
`;

exports[`Hero block renders a hero component 5`] = `
<div
  className="eea hero-block inverted full-height"
>
  <div
    className="hero-block-image-wrapper"
  >
    <div
      className="hero-block-image"
      style={
        Object {
          "backgroundImage": "url(/foo/bar.gif/@@images/image)",
        }
      }
    />
    <div
      className="hero-block-image-overlay dark-overlay-4"
    />
  </div>
  <div
    className="hero-block-inner-wrapper d-flex flex-items-center"
  >
    <div
      className="hero-block-body"
    >
      <div
        className="hero-block-text color-fg-white text-left"
      >
        <div
          className="quoted-wrapper"
        >
          Text test
        </div>
      </div>
      <div
        className="hero-block-meta text-left"
      >
        Test meta
      </div>
    </div>
  </div>
</div>
`;

exports[`Hero block renders a hero component 6`] = `
<div
  className="eea hero-block inverted full-height"
>
  <div
    className="hero-block-image-wrapper full-width"
  >
    <div
      className="hero-block-image"
      style={
        Object {
          "backgroundImage": "url(/foo/bar.gif/@@images/image)",
        }
      }
    />
    <div
      className="hero-block-image-overlay dark-overlay-4"
    />
  </div>
  <div
    className="hero-block-inner-wrapper d-flex flex-items-center"
  >
    <div
      className="hero-block-body"
    >
      <div
        className="hero-block-text color-fg-white text-left"
      >
        <div
          className="quoted-wrapper"
        >
          Text test
        </div>
      </div>
      <div
        className="hero-block-meta text-left"
      >
        Test meta
      </div>
    </div>
  </div>
</div>
`;

exports[`Hero block renders a hero component 7`] = `
<div
  className="eea hero-block inverted full-height"
>
  <div
    className="hero-block-image-wrapper full-width"
  >
    <div
      className="hero-block-image"
      style={
        Object {
          "backgroundImage": "url(/foo/bar.gif/@@images/image)",
        }
      }
    />
    <div
      className="hero-block-image-overlay dark-overlay-4"
    />
  </div>
  <div
    className="hero-block-inner-wrapper d-flex flex-items-center"
  >
    <div
      className="hero-block-body"
    >
      <div
        className="hero-block-text color-fg-white text-left"
      >
        <div
          className="quoted-wrapper"
        >
          Text test
        </div>
      </div>
      <div
        className="hero-block-meta text-left"
      >
        Test meta
      </div>
    </div>
  </div>
</div>
`;
